narwhal/types/src/primary.rs

```diff
@@ -865,78 +865,107 @@ impl PartialEq for Vote {
 #[enum_dispatch(CertificateAPI)]
 pub enum Certificate {
     V1(CertificateV1),
-}
-
-// TODO: Revisit if we should not impl Default for Certificate
-impl Default for Certificate {
-    fn default() -> Self {
-        Self::V1(CertificateV1::default())
-    }
+    V2(CertificateV2),
 }
 
 impl Certificate {
-    // TODO: Add version number and match on that
-    pub fn genesis(committee: &Committee) -> Vec<Self> {
-        CertificateV1::genesis(committee)
-            .into_iter()
-            .map(Self::V1)
-            .collect()
+    pub fn genesis(protocol_config: &ProtocolConfig, committee: &Committee) -> Vec<Self> {
+        if protocol_config.narwhal_certificate_v2() {
+            CertificateV2::genesis(committee)
+                .into_iter()
+                .map(Self::V2)
+                .collect()
+        } else {
+            CertificateV1::genesis(committee)
+                .into_iter()
+                .map(Self::V1)
+                .collect()
+        }
     }
 
     pub fn new_unverified(
+        protocol_config: &ProtocolConfig,
         committee: &Committee,
         header: Header,
         votes: Vec<(AuthorityIdentifier, Signature)>,
     ) -> DagResult<Certificate> {
-        CertificateV1::new_unverified(committee, header, votes)
+        if protocol_config.narwhal_certificate_v2() {
+            CertificateV2::new_unverified(committee, header, votes)
+        } else {
+            CertificateV1::new_unverified(committee, header, votes)
+        }
     }
 
     pub fn new_unsigned(
+        protocol_config: &ProtocolConfig,
         committee: &Committee,
         header: Header,
         votes: Vec<(AuthorityIdentifier, Signature)>,
     ) -> DagResult<Certificate> {
-        CertificateV1::new_unsigned(committee, header, votes)
+        if protocol_config.narwhal_certificate_v2() {
+            CertificateV2::new_unsigned(committee, header, votes)
+        } else {
+            CertificateV1::new_unsigned(committee, header, votes)
+        }
     }
 
-    pub fn new_test_empty(author: AuthorityIdentifier) -> Self {
-        CertificateV1::new_test_empty(author)
+    pub fn new_test_empty(protocol_config: &ProtocolConfig, author: AuthorityIdentifier) -> Self {
+        if protocol_config.narwhal_certificate_v2() {
+            CertificateV2::new_test_empty(author)
+        } else {
+            CertificateV1::new_test_empty(author)
+        }
     }
 
     /// This function requires that certificate was verified against given committee
     pub fn signed_authorities(&self, committee: &Committee) -> Vec<PublicKey> {
         match self {
             Certificate::V1(certificate) => certificate.signed_authorities(committee),
+            Certificate::V2(certificate) => certificate.signed_authorities(committee),
```

Comment:
I think method implementation like
```
match self {
    Certificate::V1(certificate) => certificate.signed_authorities(committee),
    Certificate::V2(certificate) => certificate.signed_authorities(committee),
}
```
Can be avoided by having `signed_authorities()` in `CertificateAPI`, and implementing the method separately for both V1 and V2. Is this feasible?

Reply:
The signature of `signed_authorities()` look different then the signature here, so I'm not sure how it is working right now. But I believe you don't need to implement `signed_authorities()` in `impl Certificate` with a match statement, as long as the method is implement properly for V1 and V2. The match statement will be generated by the `#[enum_dispatch(CertificateAPI)]`. We can follow up in another PR.